text/0000-programmatic-style-injection.md

```diff
@@ -0,0 +1,172 @@
+---
+title: Programmatic style injection
+status: DRAFTED
+created_at: 2021-09-16
+updated_at: 2021-09-30
+pr: (leave this empty until the PR is created)
+---
+
+# Programmatic style injection
+
+## Summary
+
+This proposal describes how LWC components can programmatically associate CSS stylesheets to an LWC component.
+
+## Basic example
+
+```js
+import { LightningElement } from 'lwc';
+import myStyles from './styles.css';
+
+export default Example class extends LightningElement {
+    static stylesheets = [
+        myStyles
+    ];
+}
+```
+
+## Motivation
+
+CSS custom properties and CSS `::part()` are the standard mechanisms by which applications can influence the look and feel of shadow DOM components. Component authors are required to explicitly define those extension points that can be styled from outside the shadow tree. While this approach gives a certain level of control over the component appearance, in certain cases developers might want to take full control over the styles injected in the shadow tree. And this without compromising the component encapsulation. 
+
+Component libraries are a prime example of this use case. Component authors want to expose high-quality base components with minimal styles and let application developers add their branding.
+
+LWC implicit template and stylesheet resolution system makes it impossible today to extend from a base LWC component and override its styles at the same time. This proposal addresses this issue by exposing a mechanism to associate stylesheets directly with the LWC component class. 
+
+## Detailed design
+
+A new `stylesheets` static property is added on the `LightningElement` constructor. This property allows developers to define which stylesheets the component should use. It accepts an array of [LWC stylesheets](#lwc-stylesheets-modules). The base `LightningElement.stylesheets` static property returns an empty array.

```

Comment:
> The base `LightningElement.stylesheets` static property returns an empty array.

This does not seem possible to implement. The problem is that the `LightningElement` object is [frozen and sealed](https://github.com/salesforce/lwc/blob/06355248d748d9b14a4bfc837c08a2ec3236c542/packages/%40lwc/engine-dom/src/apis/lightning-element.ts#L49-L50).

So we cannot _both_ set it to an empty array (allow people to do `...super.stylesheets`), and also allow `static stylesheets` in the element itself. It throws an error: `Cannot assign to read only property 'stylesheets'`.

I would recommend making `LightningElement.stylesheets` undefined. And we can tell developers that using `...super.stylesheets` is fine, unless you inherit directly from `LightningElement`. This has some precedent in the fact that `super.connectedCallback()` will throw an error unless you inherit from some non-`LightningElement` class.

Reply:
> This does not seem possible to implement. The problem is that the `LightningElement` object is [frozen and sealed](https://github.com/salesforce/lwc/blob/06355248d748d9b14a4bfc837c08a2ec3236c542/packages/%40lwc/engine-dom/src/apis/lightning-element.ts#L49-L50).

I am not certain of this. The following snippet works as expected ([fiddle](https://jsfiddle.net/g6j8kqhr/)):

```js
'use strict';

class Base {
  static stylesheets = [];
}

Object.freeze(Base);
Object.seal(Base.prototype);

class Foo extends Base {
  static stylesheets = ['foo'];
}

console.log(Base.stylesheets);
console.log(Foo.stylesheets);
```

I think the issue comes from the fact that LWC component class fields are currently transformed with [loose](https://babeljs.io/docs/en/babel-plugin-proposal-class-properties#loose) set to `true` ([compiler](https://github.com/salesforce/lwc/blob/06355248d748d9b14a4bfc837c08a2ec3236c542/packages/%40lwc/compiler/src/transformers/javascript.ts#L45)). With this flag on, the class fields are transformed into assignments instead of property definitions ([playground](https://babeljs.io/repl#?browsers=defaults%2C%20not%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=false&corejs=3.6&spec=false&loose=true&code_lz=OQVwzgpgBGAuBOBLAxrYBuAUJ5AbAhmGFAEKHQDemAkHPrCjLAJ64RgAWEEsxAvFADaAXSwBfbAHkARgCsIqAHQAzeNwBeEABRlIASiwz5SyPlw7yigA7wA9rHvMrEA9jyFiAMVu2oEAB6wEAB2ACbEupQ0dAzITKzsXDz8QsDKPsCimBI4tsFgtmyKuLYA5haQinAJnNy8rsh5BUUl5d62VSxstckGQA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=false&presets=env&prettier=false&targets=&version=7.19.6&externalPlugins=&assumptions=%7B%7D)).

Reply:
@pmdartus You must be correct. I was seeing the loose transformation when I was debugging this.

Do you think it's a big problem if developers need to do:

```js
static stylesheets = [...(super.stylesheets || []), stylesheets]
```

It's ugly, but I'm not terribly bothered by it. 🤷

Reply:
It's far from being ideal. But it's OK.